framework/src/mediaplayer/mediaplayer.c

```diff
@@ -0,0 +1,527 @@
+{
+	if (m->state == MEDIA_STATE_PLAYING) {
+		m->state = MEDIA_STATE_PAUSING;
+		return MEDIA_OK;
+	}
+	return MEDIA_ERROR;
+}
+
+media_result_t media_resume(media_t *m)
+{
+	if (m->state == MEDIA_STATE_PAUSED) {
+		m->state = MEDIA_STATE_PLAYING;
+		sem_post(&g_sem_playing);
+		return MEDIA_OK;
+	}
+	return MEDIA_ERROR;
+}
+
+media_state_t media_get_state(media_t *m)
+{
+	return m->state;
+}
+
+static media_result_t read_wav_header(media_t *m)
+{
+	//struct pcm_config config;
+	struct wav_header_s header;
+	int ret;
+
+	ret = read(m->fd, &header, sizeof(header));
+	if (ret != sizeof(header)) {
+		return MEDIA_ERROR_UNKNOWN_FILE;
+	}
+
+	// Todo: Add header information to media_t
+
+	return MEDIA_OK;
+}
+
+media_type_t get_media_type(char *path, media_t * m)
+{
+	char *dot = strrchr(path, '.');
+	if (dot != NULL) {
+		if (strcmp(dot, ".wav") == 0) {
+			return MEDIA_TYPE_WAV;
+		}
+	}
+	return MEDIA_TYPE_PCM;
+}
+
+media_t *media_open(char *path, media_op_t op, media_play_type_t play_type, struct pcm_config *config)
+{
+	media_t *m = (media_t *)malloc(sizeof(media_t));
+	if (m == NULL) {
+		return NULL;
+	}
+
+	m->fd = -1;
+	m->op = op;
+	m->state = MEDIA_STATE_CREATED;
+	m->play_type = play_type;
+	m->config = config;
+	if (play_type == MEDIA_PLAY_TYPE_FILE) {
+		m->type = get_media_type(path, m);
+		if (m->type == MEDIA_TYPE_WAV || m->type == MEDIA_TYPE_PCM) {
+			if (op == MEDIA_OP_PLAYBACK) {
+				m->fd = open(path, O_RDONLY);
+				if (m->fd < 0) {
+					goto error_out;
+				}
+
+				if (m->type == MEDIA_TYPE_WAV) {
+					read_wav_header(m);
+				}
+				m->state = MEDIA_STATE_PLAYING;
+				return m;
+			} else if (op == MEDIA_OP_RECORD) {
+				m->fd = open(path, O_RDWR | O_CREAT | O_TRUNC);
+				if (m->fd < 0) {
+					goto error_out;
+				}
+				m->state = MEDIA_STATE_RECORDING;
+				return m;
+			}
+		}
+
+	}
+error_out:
+	free(m);
+	return NULL;
+}
+
+media_result_t media_close(media_t *m)
+{
+	int ret;
+	if (m == NULL) {
+		return MEDIA_OK;
+	}
+	if (m->fd >= 0) {
+		ret = close(m->fd);
+		if (ret < 0) {
+			return MEDIA_ERROR;
+		}
+	}
+	free(m);
+	return MEDIA_OK;
+}
+
+
+int on_media_state_playing(struct pcm *pcmout, media_t *m, char *buffer, unsigned int buffer_size)
+{
+	int readed;
+	int remain;
+	int ret;
+	readed = read(m->fd, buffer, buffer_size);
+	if (readed) {
+		remain = pcm_bytes_to_frames(pcmout, readed);
+		while (remain > 0) {
+			ret = pcm_writei(pcmout, buffer + readed - pcm_frames_to_bytes(pcmout, remain), remain);
+			if (ret > 0) {
+				remain -= ret;
+			} else {
+				return ret;
+			}
+		}
+	}
+
+	return readed;
+}
+
+int on_media_state_recording(struct pcm *pcmin, media_t *m, char *buffer, unsigned int buffer_size)
+{
+	int readed;
+	int remain;
+	int ret;
+
+	readed = pcm_readi(pcmin, buffer, pcm_bytes_to_frames(pcmin, buffer_size));
+	if (readed) {
+		remain = pcm_frames_to_bytes(pcmin, readed);
+		while (remain > 0) {
+			ret = write(m->fd, buffer + pcm_frames_to_bytes(pcmin, readed) - remain, remain);
```

Comment:
There's more chance that use ram buffer instead of file system on flash than using fs for voice product.
We'd better make mechanism to send data buffer when it filled with recorded data to application layer, as I reviewed.